### PR TITLE
[SPARK-25088][CORE][MESOS][DOCS] Update Rest Server docs & defaults.

### PR DESCRIPTION
**Pull request by squito**
## What changes were proposed in this pull request?

(a) disabled rest submission server by default in standalone mode
(b) fails the standalone master if rest server enabled & authentication secret set
(c) fails the mesos cluster dispatcher if authentication secret set
(d) doc updates
(e) when submitting a standalone app, only try the rest submission first if spark.master.rest.enabled=true

otherwise you'd see a 10 second pause like
18/08/09 08:13:22 INFO RestSubmissionClient: Submitting a request to launch an application in spark://...
18/08/09 08:13:33 WARN RestSubmissionClient: Unable to connect to server spark://...

I also made sure the mesos cluster dispatcher failed with the secret enabled, though I had to do that on slightly different code as I don't have mesos native libs around.

## How was this patch tested?

I ran the tests in the mesos module & in core for org.apache.spark.deploy.*

I ran a test on a cluster with standalone master to make sure I could still start with the right configs, and would fail the right way too.

### CODE REVIEW

**Review by srowen — APPROVED** [review]

**Comment by felixcheung** [comment]
@tnachen

**Review by felixcheung — APPROVED** [review]

File: resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcher.scala [hunk 1]
```diff
@@ -51,6 +51,13 @@ private[mesos] class MesosClusterDispatcher(
     conf: SparkConf)
   extends Logging {
 
+  {
+    val authKey = SecurityManager.SPARK_AUTH_SECRET_CONF
```
Comment (tnachen) [inline resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcher.scala]: I think it might be better to place this in the MesosRestServer code, since it's not really about the framework (MesosClusterDispatcher) but the RestServer receiving requests.

**Review by tgravescs — APPROVED** [review]
lgtm. +1

File: resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcher.scala (hunk 1 above)
Comment (squito) [reply resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcher.scala]: I thought about this too, and originally wrote it that way ... but then I figured the way it works now, its really the same thing.  If I put it in the MesosRestServer, it might seem like you could run the ClusterDispatcher without the RestServer somehow -- but maybe the exception itself is clear enough?

anyway, don't feel particularly strongly either way.

Comment (tnachen) [reply resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcher.scala]: Got it, my reasoning is that it could be harder for someone looking at the code to figure out why this is not allowed, since we don't really mention about the rest server which is really the one requiring security to be turned off. Another reason it will be beneficial to have the check in the MesosRestServer is that the MesosClusterDispatcher framework could technically be decoupled from the MesosRestServer and allow another way to receive requests, so to increase flexibility and avoid someone forgetting about why we put this here, my suggestion is to move the check closer to where it's being required will help maintain this a bit better.

Comment (tgravescs) [reply resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcher.scala]: Personally the way I read this, I agree with @squito and it should be here. From my understanding from reading the docs on running on mesos, the user doesn't start the rest server, they start the dispatcher. The dispatcher doesn't support any sort of authentication.  The fact that it uses the rest server is an implementation detail the user doesn't necessarily care about.  If you change the dispatcher to have another way then it should support authentication as well or have this same error. Someone adding another way other then rest server should be made aware of this, so it being here accomplishes that.

Perhaps just adding a comment in the code about the rest server would help to clarify to developers?

Comment (tgravescs) [reply resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcher.scala]: Note if someone add another way to that supports authentication we just move it at that point.

Comment (squito) [reply resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcher.scala]: > MesosClusterDispatcher framework could technically be decoupled from the MesosRestServer

I actually think that is a reason to keep it here.  If somebody adds another way, then this check is still in place for the new way -- unless they consciously think about making it work, and then they'd move the check to a more appropriate spot.

I can add a comment in the code: `// This doesn't support authentication because the RestSubmissionServer doesn't support it.`

**Comment by tnachen** [comment]
LGTM as well

**Review by jerryshao — APPROVED** [review]
LGTM.

**Comment by tgravescs** [comment]
+1

**Comment by srowen** [comment]
Merged to master

**Comment by squito** [comment]
any objections about putting this in prior branches as well?

**Comment by felixcheung** [comment]
in this case maybe ok. perhaps just rel note this iff there's another 2.2.x or 2.1.x releases?

**Comment by akirillov** [comment]
Hi @squito, if I may, I have a few comments to share regarding this change and probably you can help me to understand it in better details.

From  [Spark Security Doc page](https://spark.apache.org/docs/latest/security.html#standalone-mode-only) it looks like `spark.master.rest.enabled` property belongs to Spark Standalone Master only. IMO, It makes sense, because it's the server side who is responsible for enabling and disabling REST endpoint according to the [source code](https://github.com/apache/spark/blob/v2.4.0/core/src/main/scala/org/apache/spark/deploy/master/Master.scala#L154-L157).

Now, from the [comment in the code] (https://github.com/apache/spark/blob/v2.4.0/core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala#L176-L180) of `SparkSubmit.scala` it looks like REST-based gateway introduced in Spark 1.3 is the default way of communication with Spark Standalone Master with a fallback to legacy RPC.

And here's the problem: now `spark.master.rest.enabled` became a client-side property, not Spark Standalone Master property. So in order to communicate with Master, a user have to provide `spark.master.rest.enabled` with each job she submits because without it, the submission will fail. This is a pretty breaking change.

What is the reason to fail the start of [MesosClusterDispatcher](https://github.com/apache/spark/blob/v2.4.0/resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcher.scala#L54-L60) if `spark.authenticate.secret` is set? Mesos uses different mechanisms for authentication and there's no need to fail the Dispatcher - this property can be just ignored or logged at most.

TL;DR It looks like `spark.master.rest.enabled` should reside only in Spark Standalone Master code and shouldn't leak into the client side of `SparkSubmit`. Also, it can be completely ignored with Mesos Dispatcher, because REST interface is the only endpoint it exposes.

Please let me know what you think. Thanks.
